core/src/block_sync.rs

```diff
@@ -74,6 +76,23 @@ impl BlockSynchronizer {
 
     /// Sends request for latest blocks to a random peer
     async fn request_block(&mut self) {
+        let now_height = {
+            let state_view = self.state.view();
+            state_view.height()
+        } as u64;
+        if now_height == self.last_sumeragi_height {
+            self.seen_blocks.clear();
+        }
+        self.last_sumeragi_height = now_height;
```

Comment:
```suggestion
        let now_height = self.state.view().height();
```

this looks like a micro optimization to me. Since you're later clearing `seen_blocks` anyways

Reply:
No it's actually essential. If the block height is not changing we need to stop filtering the incoming blocks in case we need to receive the same block again. I added a comment.

Reply:
I see. This handles soft fork. Another approach would be to always request the latest block again, wouldn't it?

Reply:
Yes. But they should be rare and so we don't want to send a wasted block on each block sync.